rpc/resource/uuid/resource_test.go

```diff
@@ -0,0 +1,240 @@
+		if (err != nil && tc.ExpectedError != err.Error()) || (err == nil && tc.ExpectedError != "") {
+			t.Errorf("tc %d:invalid error message %q, expected %q", n, err, tc.ExpectedError)
+		}
+		if id == nil {
+			continue
+		}
+		if id.Valid() != tc.Valid {
+			t.Errorf("tc %d: resource valid is %t, expected %t", n, id.Valid(), tc.Valid)
+		}
+		if id.External() != tc.External {
+			t.Errorf("tc %d: resource fq is %t, expected %t", n, id.External(), tc.External)
+		}
+		if id.ApplicationName() != tc.ApplicationName {
+			t.Errorf("tc %d: invalid application name %s, expected %s", n, id.ApplicationName(), tc.ApplicationName)
+		}
+		if id.ResourceType() != tc.ResourceType {
+			t.Errorf("tc %d: invalid resource type %s, expected %s", n, id.ResourceType(), tc.ResourceType)
+		}
+
+		if tc.NotEmpty {
+			if id.ResourceID() == "" {
+				t.Errorf("invalid resource id %s, expected non empty", id.ResourceID())
+			}
+			continue
+		}
+
+		if id.ResourceID() != tc.ResourceID {
+			t.Errorf("tc %d: invalid resource id %s, expected %s", n, id.ResourceID(), tc.ResourceID)
+		}
+	}
+}
+
+func TestCodec_Encode(t *testing.T) {
+	codec := NewCodec("app", "res")
+
+	tcases := []struct {
+		ID              resource.Identifier
+		ApplicationName string
+		ResourceType    string
+		ResourceID      string
+		ExpectedError   string
+	}{
+		{
+			ID: nil,
+		},
+		{
+			ID: (*identifier)(nil),
+		},
+		{
+			ID: &identifier{},
+		},
+		{
+			ID: &identifier{
+				resourceID: "uuid",
+				valid:      true,
+			},
+			ExpectedError: "uuid: invalid resource id uuid in uuid codec of app/res - invalid UUID length: 4",
+		},
+		{
+			ID: &identifier{
+				resourceID: "cd89d33a-7310-11e8-8297-787b8ac3ce32",
+				valid:      true,
+			},
+			ApplicationName: "app",
+			ResourceType:    "res",
+			ResourceID:      "cd89d33a-7310-11e8-8297-787b8ac3ce32",
+		},
+	}
+
+	for n, tc := range tcases {
+		pb, err := codec.Encode(tc.ID)
+		if (err != nil && tc.ExpectedError != err.Error()) || (err == nil && tc.ExpectedError != "") {
+			t.Errorf("tc %d:invalid error message %q, expected %q", n, err, tc.ExpectedError)
+		}
+		if pb == nil {
+			continue
+		}
+
+		if v := pb.GetApplicationName(); v != tc.ApplicationName {
+			t.Errorf("tc %d: invalid application name %s, expected %s", n, v, tc.ApplicationName)
+		}
+		if v := pb.GetResourceType(); v != tc.ResourceType {
+			t.Errorf("tc %d: invalid resource type %s, expected %s", n, v, tc.ResourceType)
+		}
+		if v := pb.GetResourceId(); v != tc.ResourceID {
+			t.Errorf("tc %d: invalid resource id %s, expected %s", n, v, tc.ResourceID)
+		}
+	}
+}
+
+func TestIdentifier_Scan(t *testing.T) {
+	tcases := []struct {
+		Value           interface{}
+		Valid           bool
+		External        bool
+		ApplicationName string
+		ResourceType    string
+		ResourceID      string
+		ExpectedError   string
+	}{
+		{
+			Value:      "uuid",
+			Valid:      true,
+			ResourceID: "uuid",
+		},
+		{
+			Value: nil,
+		},
+		{
+			Value:         2018,
```

Comment:
So to add resource identifier support for incremental PK one should implement Identifier and Codec separately, right?

Reply:
Right!